File: Behavioral/template-method/examples/name-generator/Python/NameGenerator/generators/EnglishNameGenerator.py
from generators.BaseNameGenerator import BaseNameGenerator, Gender
import random


class EnglishNameGenerator(BaseNameGenerator):
    __first_male_names = ["Jim", "Richard", "Ted", "Robert", "Danny"]
    __first_female_names = ["Kate", "Susan", "Victoria", "Diana"]
    __surnames = ["Smith", "Black", "Robinson", "Rand"]

    def generate_first_name(self, gender: Gender) -> str:
        if gender == Gender.MALE:
            return self.__first_male_names[random.randint(0, len(self.__first_male_names)-1)]
        return self.__first_female_names[random.randint(0, len(self.__first_female_names)-1)]

    def generate_surname(self, gender: Gender) -> str:
        return self.__surnames[random.randint(0, len(self.__surnames)-1)]

File: Behavioral/template-method/examples/name-generator/C#/NameGenerator/NameGenerator/Generators/EnglishNameGenerator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace NameGenerator.Generators
{
    public class EnglishNameGenerator: BaseNameGenerator
    {
        private readonly string[] _firstMaleNames = new string[] { "Jim", "Richard", "Ted", "Robert", "Danny" };
        private readonly string[] _firstFemaleNames = new string[] { "Kate", "Susan", "Victoria", "Diana" };
        private readonly string[] _secondNames = new string[] { "Smith", "Black", "Robinson", "Rand" };

        protected override string GenerateFirstName(Gender gender)
        {
            switch (gender)
            {
                case Gender.Male:
                    return _firstMaleNames[_random.Next(0, _firstMaleNames.Length)];
                case Gender.Female:
                    return _firstFemaleNames[_random.Next(0, _firstFemaleNames.Length)];
                default:
                    throw new ArgumentOutOfRangeException();
            }

        }

        protected override string GenerateSurname(Gender gender)
        {
            return _secondNames[_random.Next(0, _secondNames.Length)];
        }
    }
}
